src/server/main_service.cc

```diff
@@ -986,6 +987,12 @@ bool Service::InvokeCmd(const CommandId* cid, CmdArgList tail_args, ConnectionCo
   DCHECK(cid);
   DCHECK(!cid->Validate(tail_args));
 
+  if (!acl::IsUserAllowedToInvokeCommand(*cntx, *cid)) {
+    (*cntx)->SendError(absl::StrCat("User: ", cntx->authed_username,
+                                    " does not have the credentials to execute this command"));
+    return true;
+  }
```

Comment:
Here's the catch and it's actually a very dangerous bug. I added specifically the Verify functions but forgot to add a comment here: Any validation logic should only be placed in verify functions. 

The reason for is that verification is always performed on a const connection context pointer. Its possible to do this in parallel. Executing in parallel with one context is not possible, so I use just "stub" contexts to transfer replies. 

So first:

Please add a test case with squashing. You can do the following: 1. enable `multi_exec_squash` 2. Execute a 10 command multi/exec transaction that spans multiple shards (important - use simple single shard commands). 

(IDK why you actually use pytests for everything. They're easier to use, sure 🙂 But they don't run on every pr)

Second: 

There are two solutions:

1. Use VerifyCommand state to check allowance. It has docs in its header above it. It's fine for all cases except MULTI/EXEC, because with it, commands are verified **only** when added to the multi/exec buffer. So if at the moment EXEC is called the acl rights change, the verification is invalid. 

2.  See `DispatchMonitor` below. It checks conn_context.squashing_info.parent? Smth like this, I don't remember 🙂 This way you can get a const pointer to the original context. 

3. Copy acl_username into the stub context, but I'm not a fan of this... Better let validation operate on the original context
 
Sorry for the inconvenience with my squashing stuff 😅 😄

Reply:
Sorry for my late reply, I wanted to go over in detail and understand what the exact issue is and what have I missed. Even though you explained, I originally thought that I had it covered when I first pushed this PR and added the error checking within the `InvokeCommand` but what I didn't take into account is as you mentioned the `stub` context. The good thing is that I learned a few things and I now have a better understanding of our squashing/multi transaction mechanism.

So a few things:

1. We can't really rely only `VerifyCommandState` or on `VerifyCommandExecution`. We need both and the reason is that for `multi` transactions we need to check twice as per your first point. a) when we issue the command, it could be the case, that the user does not have the credentials and therefore we can issue an error immediately. b) a `multi` command was issued and some admin changed the user's acl categories between the call of `multi` and the call to `exec`. For this particular case we need to use `VerifyCommandExecution` which covers that corner case. Best part is that we now also have two different errors. For the first case, the error is that the user does not have the credentials and for the later that the user's acl categories have changed inbeteween. IMO this is far clearer for the user + this is what redis does.

2. There is an issue that we send multiple error responses, I will fix this but on another PR (plz see my comment on the pytest -- it will become clear what the problem is).

> Copy acl_username into the stub context, but I'm not a fan of this... Better let validation operate on the original context

Neither am I, we should perform checks on the OG context and that's what my changes do.

> (IDK why you actually use pytests for everything. They're easier to use, sure 🙂 But they don't run on every pr)

Because the most important for now is to have to end to end, I don't want to change unit tests until I have something stable. I will am planning to add them.

P.s. I am not a big fun of having the regression tests not being run per PR but I understand the rationale of why we wouldn't want that. I think this will be problematic especially when the team grows as we will start having cascading failures (multiple people who pushed in a 3 hour window can all break the regression tests) but for now this is a non-issue so it's not worth discussiing